### DISCUSSION TITLE
Allow left joins with custom tables join

### DISCUSSION THREAD

**Discussion by QuestionAndAnswer** [mikro-orm/mikro-orm#3580] [Ideas]
**Is your feature request related to a problem? Please describe.**

So, I'm not using many-to-many relation provided by MikroOrm via decorators, as it the common issue in TS, like in Nest for example - it breaks compile time validations, and it in case of MikroOrm, collections also prevents me from writing pure functions for data mapping, as I would need to first create an object, and provide into `new Collection` correct `this`, and then I'd need to populate it.

Such trick is not catched by TS also, (like did I initialized collection or not). More over proceeding work with collection also leaves me in a situation where I need to constantly check is collection initialized or not.

It quickly becoming messy, and less stable because of lack verifiability property of such approach.

Also, it often creates circular dependencies and blown up entities definition, which is hard to work with. Native SQL way of modeling entities and relationships just feels more native and easier in this way.

There are another couple of reasons, but the main one - decorators and runtime linking and checks leaks more errors into runtime, instead of compile time. And I'd like to have as much errors as possible on compile time. Thanks to mikro orm TS-morph it's already providing a lot, but not enough with many to many and collections (while generally it's good, but just not my way of developing).

Assume instead I doing many to many relations manually.

```typescript
@Entity()
class Rel {
  @ManyToOne({ primary: true, cascade: [] })
  d1: IdentifiedReference<A>;
  @ManyToOne({ primary: true, cascade: [] })
  d2: IdentifiedReference<B>;
}

@Entity()
class A {
  @PrimaryKey()
  id: string;
  @Property()
  name: string;
}

@Entity()
class B {
  @PrimaryKey()
  id: string;
  @Property()
  name: string;
}
```

And I'd like to be able to filter out A's and B's which has a references.
In this case I need a way to provide custom left join into query builder, in order to be able to filter out A's which has B's referencing it and vice versa.

Right now I'm doing something like this.

```typescript
export function withInUseSearch<TEntity extends AnyEntity>(
    relEntityType: Constructor<AnyEntity>,
    entityType: Constructor<TEntity>,
    em: EntityManager,
) {
    const relMeta = em.getMetadata().get(relEntityType.name);
    const rel = relMeta.relations.find(x => x.type === entityType.name);

    assert(rel, `couldn't resolve relation ${relEntityType.name} <-> ${entityType.name}`);

    return (qb: SelectQueryBuilder<TEntity>, inUse: boolean | null | undefined) => {
        const kqb = qb.getKnexQuery();

        if (typeof inUse !== 'boolean') return kqb;

        kqb.leftJoin(`${relMeta.tableName} as rel`, `${qb.alias}.id`, `rel.${rel.name}_id`);

        if (inUse) {
            kqb.whereNotNull('product_id');
        } else {
            kqb.whereNull('product_id');
        }

        return kqb;
    };
}

const addInUseFilter = withInUseSearch(Rel, A, B, em);

async function find(filter: Filter) {

  const qb = em.createQueryBuilder(A)
    .select("*")
    .where(somefields here which are part of A definition);

  const kqb = addInUseFilter(qb, filter.inUse);

  const rawEntities = await this.em.execute(kqb);
  const entities = rawEntities.map(x => this.em.map(A, x));

  return entities;
}
```

So, basically I'm attaching raw query via knex.

**Describe the solution you'd like**
What I was thinking is:

Add a possibility into `qb.leftJoin*` methods to provide custom join statements, as left join by it's nature will not destruct entity for which this `qb` was created, only additional fields will be added which might be used for filter purpose only. qb should be able to extract target entity without any issues always. My example is working at least as a first proof for that.

So it would be something like this:
```typescript
const qb = em.createQueryBuilder(A)
  .select("*")
  .where(somefields here which are part of A definition)
  // 1 - another entity, 2 - any IdentifiedReference field from 1, 3 - any field from target entity (A in this case)
  .leftJoin(Rel, `id`, `d1`)
  .where('d1', { $ne: null });
const entities = await qb.execute();
return entities;
```

**Describe alternatives you've considered**
In order to make such joins type safe we might consider adding typed string literals for custom queries. It's just a thought, and I'm not quite sure how to do it without extended code generation procedures. But, it should be doable as https://jawj.github.io/zapatos/ already have such functionality. Of course zapatos doing it from scratch, but MikroORM should be able to extract same amount of information required for that as well after parsing Entities and forming metadata files for cache.

This alternative solution is big, but unleashes the same amount of functionality and zero cost abstractions in some places as zapatos does. This might be considered for a major release if you like.

**Additional context**

**Comment by B4nan** [comment]
I am not sure I understand your motivation. This sounds like a duplicate of #593, I dont think it make sense to allow overriding things on QB level, this should be rather about modifying entity definition. Only that way we can have it type safe.

FYI `qb.execute()` will give you raw data, not entities.

edit: Moving to discussions before we settle on what you are actually asking for. Checked your snippet once again, I dont see anything that couldnt be done currently. Lets start with providing complete repro for what you do, and the queries you are trying to generate. If you are asking for a way to do any query, regardless of the entity metadata (which is kinda your example too), that is out of scope, use knex for such. The ORM QB needs to work based on the metadata only so it knows how to hydrate the results, that is its purpose, you have knex for anything else.

**Comment by QuestionAndAnswer** [comment]
So, I'll try to rephrase what is the issue.

1. I don't want to use m-to-m relation provided via decorator as this creates circular references in entities -> potentially leads to mikro orm inability for entities auto discovery and nasty bugs in project build.
2. I don't want to use m-to-m relation provided via decorator as this requires me usage of `Collection` type, which is good for type safety, but requires from me to define constructor on my entities with boilerplate of fields initialization -> all of it leads to inability creating pure functions for child creation and population in one hit. Like

```typescript
@Entity()
class A {
  @PrimaryKey()
  id: string;
  // btw, this is might be issue with my setup, but some why Collection doesn't recognize `this` in constructor later on
  // so I need to pass here this explicitly
  @ManyToMany(...)
  bs = new Collection<B, this>(this);
  
  // more fields
  
  // here I would like to have exactly A, as in this case I will be forced by compiler to populate all the 
  // required fields of A. But I can't pass bs field, as this is a Collection type,
  // and I can't populate Collection type externally, as it requires this on the time of creation.
  constructor (e: Omit<A, "bs">) {
    this.id = e.id;
    
    // more fields population boiler plate
  }
}

@Entity()
class B {
  @PrimaryKey()
  id: string;
  @ManyToMany(...)
  "as" = new Collection<A, this>(this);
  
  // .. the same population code as in A
}

// later on there are external entities creation / population functions

function createAVariant1(data: A) {
  const e = new A({ id: uuid(), /* more fields */});
  
  // I need to populate bs references as a separate step without compile help in terms of a is data in collection set or not
  for(let bId of data.bsIds) {
    e.add(Reference.createFromPk(B, bId));
  }
  
  return e;
}

// later loading the collection is also kind of a unverifiable by compiler problem, is it initialized or not. So it becoming a runtime error.
```

Instead I'd prefer to make hierarchical auto create embedded into call chain. But collections are not allowing me to do so. Therefore I need to model m-to-m relation manually, and then I can create all of them pure.

```typescript
@Entity()
class A {
  @PrimaryKey()
  id: string;
}

@Entity()
class B {
  @PrimaryKey()
  id: string
}

@Entity()
class Rel {
  @ManyToOne({ primary: true, cascade: [] })
  a: IdentifiedReference<A>;
  @ManyToOne({ primary: true, cascade: [] })
  b: IdentifiedReference<B>;
}

// of course the need in such functions might as well gone in this scenario
function createA(data) {
  return em.create(A, data);
// note usage of constructor is perfectly fine here also, as long as there are no need for fields modification after it.
}
function createB(data) {
  return em.create(B, data);
}
function createRel(a, b) {
  // don't remeber exaxct syntax, but basically passing a refs here.
  // constructor for this particular case is easier and preferable
  // but still no need to modify fields after create.
  return em.create(Rel, { a: a, b: b });
}
```

So with the latest there are no codependency between A and B, no circular dependencies, and referencing is of course exist if Rel entity populated. I get compile time checks on required fields cheaper (I don't need to define constructors which I need). I'm more explicit in what I'm doing and closer to DB, which gives me a bit more space for optimization. I don't have errors of loaded collection not being initialized in runtime, but I'll have no possibility to work with relation without explicitly resolving relation entity first, which even prevents any compile time errors as there is no possible logical code for that written yet.

And then, without fields defined in the entities itself I need a way still to search relative to aggregates A or B, if they have a ref while still searching by own filter criterias of A or B.
So, the only way to do it in one DB call - SQL joins (for SQL of course).
Joins currently only possible via syntax `leftJoin(field, alias, cond)`, which tightly binds it to owning entity. But, as left join is additive operator, and doesn't destructs initial selects search space, it might be extended for broader form `leftJoin(foreignEntity, foreignEntityField, ownField, condition?)`. 

I need this second form to be able now to filter A with some name and which has relations to B's for example. With current API available I can do it only via knexQueryBuilder and lose of all typesafety. The second proposed call form still validatable, as `qb` created requires root entity to be specified (A) and on a call of `leftJoin` `foreignEntityFields` might be proposed out of foreignEntity param and generics of TS.

So, you might say then, that further where conditions wouldn't have info on foreignEntity, then I'd propose that it might be done as return of subQueryBuilder with extended context. Like we already have `SelectQueryBuilder`, we might introduce `JoinQueryBuilder<A, B, Rel>`, for example. This would be exactly enough information to embedded typesafety for further `qb` calls.

So, I hope this might clarify things a bit.

**Reply by B4nan** [reply]
> I don't want to use m-to-m relation provided via decorator as this creates circular references in entities -> potentially leads to mikro orm inability for entities auto discovery and nasty bugs in project build.

What do you mean by that? Circular relations are working just fine, both discovery as well as runtime usage. Its a core concept here. Also keep in mind that the entity constructor is used only for new entities - not for those loaded via EM.

> I don't want to use m-to-m relation provided via decorator as this requires me usage of Collection type, which is good for type safety, but requires from me to define constructor on my entities with boilerplate of fields initialization -> all of it leads to inability creating pure functions for child creation and population in one hit. 

Well, then we are not talking about entities at all. Just use knex if you dont want entity instances. If you want them, you will need to use the `Collection` wrapper.

If you won't use the `Collection` wrapper, nothing will work with regard to change set computation and flushing. I dont plan to relax this condition anyhow, the collection wrapper is the go to way, if you dont like it, better to find some other tool.

**Reply by QuestionAndAnswer** [reply]
> If you won't use the Collection wrapper, nothing will work with regard to change set computation and flushing.

So, that's the point, I'm not using `Collections` because I'm not using `OneToMany` or `ManyToMany` relations defining entities. I just designing entities the same way as it is done in SQL. I'm not abstracting out essential way how SQL stores such things.

>  I dont plan to relax this condition anyhow, the collection wrapper is the go to way, if you dont like it, better to find some other tool.

No, no. I'm not saying anything about `Collection` to be done. The opposite, to extend possibilities to work without them.

> What do you mean by that? Circular relations are working just fine, both discovery as well as runtime usage. Its a core concept here. Also keep in mind that the entity constructor is used only for new entities - not for those loaded via EM.

They are, but as it was said in documentation, there might appear some problems on discovery stage because of it, and alternative is to provide the list of entities manually. And not to play any kind of roulette, I'm just choosing not to use `Collections` at all.

In mikro orm workflow built around entities. That's okay. That's a logical thing to do. To implement some behavior across entities we need to use relations. That's also okay, but comes with tradeoffs, like `Collections` and whole decorators thing in JS has.

So, in this case, not using the relations way of doing cross entities operations, I need some way to do it with other abstractions let's say. One of it is `JoinQueryBuilder<A, B, Rel>`, which I'm exactly proposing. Because from my practice and experience (this sentence is biased of course), only Parent to child relations, with child lifespan limited fully by parent are only suitable for embedding into field, the rest, like graph modelling, where I need to model exactly relations as separate entities without node's being aware about them, I kind can't use field based approach.

I'll try to restate again, because I see that I'm not that clear on what I'm asking.
I want a way to work with SQL in TS, particulary through MikroORM as it has already a lot of grate things to use, to work with somehow related (let's call it special complex relation logic) SQL tables (or documents in Mongo) in a typesafe way, without being forced to model such relations only through parent child fields, but via separate relation table and absence of relation fields in a related tables (A, B and rel entities example from above).
If in current approach I always has to do `A <-> B`, and MikroORM will unwrap it into `A <- ABRel -> B`, and I'm aimed to define `A <- ABRel -> B` from scratch and work with them natively. ManyToMany in current MikroORM docs standing that I might have custom relation table defined, but still this requires in A and B to have fields relate to each other. With a JoinedQueryBuilder type this should be reachable and typesafe.
The basic difference in a way framework is forcing to write a code and modeling, while it has almost everything not to do so. Thanks that it design a 'hackable' in a good sense way.

I'm sorry If this is not clear again, I'm trying my best.)

**Reply by B4nan** [reply]
Ok I hope I have a better understanding now. I dont mind making things more extensible/configurable/flexible, I just didnt like the idea of QB allowing things that the entity definition is not reflecting (as then the `qb.getResult()` would be broken/malfunctioning), but this is clearly a separate problem. 

Regarding the M:N, you can already define it the way you want, with only m:1 relations on the pivot entity (basically just define the pivot entity). But the counterpart (and probably the thing you are up to) is the querying of it, from the leafs (where the 1:m relations to the pivot entity would be), and that is not really possible without defining the inverse sides.

I will probably need a bit more time to digest this, if you can propose more concrete ways how to go about this, PR welcome (can be just a draft/PoC/RFC).

> there might appear some problems on discovery stage because of it

That note in the docs is rather outdated, I am not aware of anyone having that problem in last two years or even more. Also today you can define the metadata without decorators (e.g via `EntitySchema`, or with a custom metadata provider). There are still some issues with ESM projects but those can be also gotten around easily.

**Reply by QuestionAndAnswer** [reply]
So, just a quick update on draft proposal syntax.

```typescript

@Entity()
class A {
    @PrimaryKey()
    id: string;

    @Property()
    nameA: string;
}

@Entity()
class B {
    @PrimaryKey()
    id: string;

    @Property()
    nameB: string;
}

@Entity()
class Rel {
    @ManyToOne({ primary: true, cascade: [] })
    a: IdentifiedReference<A>;
    @ManyToOne({ primary: true, cascade: [] })
    b: IdentifiedReference<B>;
}

function example(em: EntityManager) {
    const qb = em
        .createQueryBuilder(leftJoin(A, Rel, 'id', 'a'), leftJoin(B, Rel, 'id', 'b'))
        .select('*')
        .where({
            nameA: 'test_A',
            nameB: 'test_B',
        });

    return qb.execute(); // will return only A, as initial join type
}
```

That is what I'm thinking about at the moment.
There might be variations via standalone .leftJoin calls, but I didn't yet come up with good way to design it. As I see know, I can propose probably generic way of left join cascaded calls.

**Reply by QuestionAndAnswer** [reply]
While thinking about how to solve it, I realized that I don't see a typesafe way to make a custom joins (let's consider at least left for beginning) by fields which are not part of explicit relation. Knex QB is a non typesafe option here.
In general, SQL allows you to do joins by any field. It does not imply those fields has a dependency in the other table.
Answering this generic question might help to solve what I'm looking for. 

Am I missing something? Is it possible in MikroORM?

I'm trying to figure out how query builder works, and I'm surprised in a bad way that it is so monolithic. Quite hard to get through chain of if's when it is not a separate classes, or functions and a track huge state context. I do understand that that's how things worked out probably quicker or better, just an observation.

**Reply by QuestionAndAnswer** [reply]
And BTW, latest example of syntax I try to figure out how to implement is:

```typescript
const qb = em
            .createQueryBuilder(A)
            .select('*')
            .leftJoin(Rel, 'id', 'a') // + optional condition statement between args, or even complex expression
            .leftJoin(B, 'id', 'b')
            .where({
                nameA: 'test_A',
                nameB: 'test_B',
            });
            
const entities = qb.execute(); // will return only A instances, as it's fields are not "corrupted" by the left join itself.
```

**Reply by B4nan** [reply]
Not really supported, but it would make sense to do it and it could be rather simple. I am currently refactoring how the joins are implemented (to have more control), so I will try to incorporate this right ahead. You won't get the propsed API, the aliases will need to be used (so `where({ 'a.nameA': '...' })`) but I guess that's understandable. If there is no alias, it means it refers to the root entity.

> I'm trying to figure out how query builder works, and I'm surprised in a bad way that it is so monolithic

How would you improve it? The QB was never meant to be universal, not a knex replacement if you want. It's metadata aware and works with relations natively, that is its purpose. Sure we can make it better.

**Reply by QuestionAndAnswer** [reply]
> How would you improve it? The QB was never meant to be universal, not a knex replacement if you want. It's metadata aware and works with relations natively, that is its purpose. Sure we can make it better.

So, my idea, following yours with SelectQueryBuilder specialized interfaces, would be extract standalone classes which provide set of methods, like (SelectQueryBuilder, UpdateQueryBuilder, Delete and etc.), which would operate then on a shared query state if you'd like. With such appraoch, after custom left join overloaded call I proposed, we'd return CustomJoinQueryBuilder which would limit later on operations only for read purposes. The beauty of this would come in transitioning between such sub builders, as each other would narrow or in some cases, like with joins, extend set of operations or parameters to be passed, without modification of response in some cases. Like, you can transition from 

GenericQB -> ReadQB | WriteQB
ReadQB <-> LeftJoinQB or just generic JoinQB (extended where set)
WriteQB -> UpdateQB | DeleteQB

Plus of this approach is simplicity of embedding inner subqueries, as subquery would have different query state store, which might just merged in into higher level query state store (of course with aliases special remapping to avoid collisions, but it is basically a concept of register virtual machine variables allocation).

Another approach would be AST sql, but this one is big, and already might be considered as another library or project )

For now it is quite verbose, but I can make a PoC of this if you'd like, or you think it might be helpfull.
